velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -62,57 +67,355 @@ void generateJsonTyped(
+    }
+  });
+}
+
+void castToJsonFromMap(
+    const BaseVector& input,
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult) {
+  // input is guaranteed to be in flat encoding when passed in.
+  auto inputMap = input.as<MapVector>();
+
+  auto mapKeys = inputMap->mapKeys();
+  VELOX_CHECK_EQ(
+      mapKeys->typeKind(),
+      TypeKind::VARCHAR,
+      "The type of map keys must be VARCHAR.");
+
+  auto elementsRows = functions::toElementRows(mapKeys->size(), rows, inputMap);
+  exec::LocalDecodedVector keysHolder(context, *mapKeys, elementsRows);
+  auto keysDecoded = keysHolder.get();
+
+  exec::LocalSelectivityVector keysBaseRows(
+      context->execCtx(), keysDecoded->base()->size());
+  keysBaseRows->clearAll();
+  elementsRows.applyToSelected(
+      [&](auto row) { keysBaseRows->setValid(keysDecoded->index(row), true); });
+  keysBaseRows->updateBounds();
+
+  auto mapValues = inputMap->mapValues();
+  exec::LocalDecodedVector valuesHolder(context, *mapValues, elementsRows);
+  auto valuesDecoded = valuesHolder.get();
+
+  exec::LocalSelectivityVector valuesBaseRows(
+      context->execCtx(), valuesDecoded->base()->size());
+  valuesBaseRows->clearAll();
+  elementsRows.applyToSelected([&](auto row) {
+    valuesBaseRows->setValid(valuesDecoded->index(row), true);
+  });
+  valuesBaseRows->updateBounds();
+
+  VectorPtr keysResult;
+  BaseVector::ensureWritable(
+      *keysBaseRows, JSON(), context->pool(), &keysResult);
+  auto keysStringResult = keysResult->as<FlatVector<StringView>>();
+
+  // Map keys must not be null.
+  castToJson<TypeKind::VARCHAR, true>(
+      *keysDecoded->base(), context, *keysBaseRows, *keysStringResult);
+
+  VectorPtr valuesResult;
+  BaseVector::ensureWritable(
+      *valuesBaseRows, JSON(), context->pool(), &valuesResult);
+  auto valuesStringResult = valuesResult->as<FlatVector<StringView>>();
+
+  VELOX_DYNAMIC_TYPE_DISPATCH(
+      castToJson,
+      mapValues->typeKind(),
+      *valuesDecoded->base(),
+      context,
+      *valuesBaseRows,
+      *valuesStringResult);
+
+  // Construct the Json string of each map from Json strings of its keys and
+  // values.
+  rows.applyToSelected([&](auto row) {
+    auto offset = inputMap->offsetAt(row);
+    auto size = inputMap->sizeAt(row);
+
+    if (size == 0) {
+      flatResult.set(row, "{}");
+    } else {
+      auto keysStringSize = getSumOfElementsStringSize(
+          *keysStringResult, keysDecoded->indices(), offset, size);
+      auto valuesStringSize = getSumOfElementsStringSize(
+          *valuesStringResult, valuesDecoded->indices(), offset, size);
+
+      auto proxy = exec::StringProxy<FlatVector<StringView>>(&flatResult, row);
 
-    flatResult.set(row, StringView{result});
+      // Reserves space for Json strings of all keys and values of this map,
+      // semicolons, commas, and the enclosing curly braces.
+      proxy.reserve(keysStringSize + valuesStringSize + size * 2 + 1);
```

Comment:
Can we reserve once across all values, instead of once per array?

Reply:
Refactored to call `flatResult.getBufferWithSpace()` before per-row loops. Thank you for the suggestion!